File: scripts/validate-data.ts
import Ajv from 'ajv';
import addFormats from 'ajv-formats';
import * as fs from 'fs';
import * as path from 'path';

// ---------------------------------------------------------------------------
// Schema loading and AJV setup
// ---------------------------------------------------------------------------

const root = path.resolve(__dirname, '..');

const granthaSchemaPath  = path.join(root, 'grantha.schema.json');
const envelopeSchemaPath = path.join(root, 'grantha-envelope.schema.json');
const partSchemaPath     = path.join(root, 'grantha-part.schema.json');

const granthaSchema  = JSON.parse(fs.readFileSync(granthaSchemaPath, 'utf-8'));
const envelopeSchema = JSON.parse(fs.readFileSync(envelopeSchemaPath, 'utf-8'));
const partSchema     = JSON.parse(fs.readFileSync(partSchemaPath, 'utf-8'));

// Register schemas by the URI used in cross-file $refs so they resolve correctly.
const ajv = new Ajv({ strict: false, allErrors: true });
addFormats(ajv);
ajv.addSchema(granthaSchema,  'grantha.schema.json');
ajv.addSchema(envelopeSchema, 'grantha-envelope.schema.json');
ajv.addSchema(partSchema,     'grantha-part.schema.json');

const validateGrantha     = ajv.getSchema<Record<string, unknown>>('grantha.schema.json')!;
const validateEnvelope    = ajv.getSchema<Record<string, unknown>>('grantha-envelope.schema.json')!;
const validatePart        = ajv.getSchema<Record<string, unknown>>('grantha-part.schema.json')!;
const validateSubEnvelope = ajv.compile(
  { $ref: 'grantha-envelope.schema.json#/definitions/edition_sub_envelope' }
);

// ---------------------------------------------------------------------------
// File classification
// ---------------------------------------------------------------------------

type FileKind =
  | 'grantha-envelope'
  | 'edition-sub-envelope'
  | 'grantha-part'
  | 'grantha'
  | null;

const VALID_KINDS = new Set<string>([
  'grantha-envelope',
  'edition-sub-envelope',
  'grantha-part',
  'grantha',
]);

/**
 * Return the file's explicit kind marker, or null to skip the file.
 * Classification is a direct lookup on data.kind — no field-presence inference.
 *
 * @param data - Parsed JSON content of the file.
 * @returns The kind string if recognised, or null to skip the file.
 */
function classifyFile(data: Record<string, unknown>): FileKind {
  const kind = data.kind as string | undefined;
  return (kind !== undefined && VALID_KINDS.has(kind)) ? kind as FileKind : null;
}

/** Return the AJV compiled validator for the given kind discriminant. */
function validatorForKind(kind: Exclude<FileKind, null>) {
  switch (kind) {
    case 'grantha-envelope':    return validateEnvelope;
    case 'edition-sub-envelope': return validateSubEnvelope;
    case 'grantha-part':        return validatePart;
    case 'grantha':             return validateGrantha;
  }
}

// ---------------------------------------------------------------------------
// Structure-consistency check (ref depth vs structure_levels depth)
// ---------------------------------------------------------------------------

function getStructureDepth(levels: unknown[]): number {
  if (!levels || levels.length === 0) return 0;
  let depth = 1;
  let current = levels[0] as Record<string, unknown>;
  while (Array.isArray(current.children) && current.children.length > 0) {
    depth++;
    current = current.children[0] as Record<string, unknown>;
  }
  return depth;
}

function checkStructureConsistency(data: Record<string, unknown>): string[] {
  const errs: string[] = [];
  const levels = data.structure_levels as unknown[] | undefined;
  if (!levels || levels.length === 0) return errs; // schema will catch missing field

  const expectedDepth = getStructureDepth(levels);
  const allPassages = [
    ...((data.prefatory_material as unknown[]) || []),
    ...((data.passages          as unknown[]) || []),
    ...((data.concluding_material as unknown[]) || []),
  ] as Array<Record<string, unknown>>;

  for (const passage of allPassages) {
    const ref = passage.ref as string;
    if (!ref) continue;
    const refParts = ref.split('.');

    if (passage.passage_type === 'main') {
      if (refParts.length !== expectedDepth) {
        errs.push(
          `[structure] Passage "${ref}" has ${refParts.length} ref segments ` +
          `but structure_levels defines ${expectedDepth} levels`
        );
      }
      for (const part of refParts) {
        if (!/^\d+$/.test(part)) {
          errs.push(`[structure] Passage ref "${ref}" contains non-numeric segment "${part}"`);
        }
      }
    }
  }

  const commentaries = data.commentaries as unknown;
  if (Array.isArray(commentaries)) {
    for (const commentary of commentaries as Array<Record<string, unknown>>) {
      for (const passage of (commentary.passages as Array<Record<string, unknown>>) || []) {
        const ref = passage.ref as string;
        if (!ref) continue;
        const refParts = ref.split('.');
        if (refParts.length !== expectedDepth) {
          errs.push(
            `[structure] Commentary passage "${ref}" has ${refParts.length} ref segments ` +
            `but structure_levels defines ${expectedDepth} levels`
          );
        }
      }
    }
  }

  return errs;
}

// ---------------------------------------------------------------------------
// Recursive file scan
// ---------------------------------------------------------------------------

function collectJsonFiles(dir: string): string[] {
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  const results: string[] = [];
  for (const entry of entries) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      results.push(...collectJsonFiles(full));
    } else if (entry.isFile() && entry.name.endsWith('.json')) {
      results.push(full);
    }
  }
  return results;
}

// ---------------------------------------------------------------------------
// Edition path-resolution check (grantha-level envelopes only)
// ---------------------------------------------------------------------------

interface EditionStubShape {
  path?: unknown;
  edition_id?: unknown;
}

/**
 * Verify that every edition stub path in a grantha-level envelope resolves to
 * an existing file or directory under the library root. Catches envelope/data
 * drift (e.g. an editions[].path pointing at a directory that no longer exists
 * after a text was re-imported).
 *
 * @param data - Parsed grantha-envelope content.
 * @param libraryRoot - Absolute path to the library directory.
 * @returns List of error messages (empty when all paths resolve).
 */
function checkEditionPathsResolve(
  data: Record<string, unknown>,
  libraryRoot: string,
): string[] {
  const editions = data.editions as EditionStubShape[] | undefined;
  if (!Array.isArray(editions)) return [];
  const errs: string[] = [];
  for (const edition of editions) {
    if (typeof edition.path !== 'string') continue;
    const full = path.join(libraryRoot, edition.path);
    if (!fs.existsSync(full)) {
      const label =
        typeof edition.edition_id === 'string'
          ? edition.edition_id
          : edition.path;
      errs.push(`[editions] "${label}" path does not resolve: ${edition.path}`);
    }
  }
  return errs;
}

// ---------------------------------------------------------------------------
// Per-block presentation checks (IDEA.md per-block presentation model)
// ---------------------------------------------------------------------------

/** Pinned classification (mirrors lib/data.ts KNOWN_PASSAGE_KINDS). Any kind
 *  found in the corpus without an entry here is a build error — a future prose
 *  leaf (e.g. "Gadya") must be classified explicitly, never silently "verse". */
const KNOWN_PASSAGE_KINDS = new Set([
  'Para',
  'Gadya',
  'Shloka',
  'Mantra',
  'Verse',
  'Sutra',
]);

interface PassageShape {
  ref?: unknown;
  passage_type?: unknown;
  kind?: unknown;
}

/**
 * Verify the per-block kind invariants for a content file (flat grantha or
 * part): every `passage_type: "main"` passage carries a classified `kind`; no
 * prefatory/concluding passage carries one.
 *
 * @param data - Parsed grantha or grantha-part content.
 * @returns List of error messages (empty when the invariants hold).
 */
function checkPassageKinds(data: Record<string, unknown>): string[] {
  const errs: string[] = [];
  for (const key of ['passages', 'prefatory_material', 'concluding_material']) {
    const arr = data[key];
    if (!Array.isArray(arr)) continue;
    for (const passage of arr as PassageShape[]) {
      const ref = typeof passage.ref === 'string' ? passage.ref : '?';
      if (passage.passage_type === 'main') {
        const kind = passage.kind;
        if (typeof kind !== 'string' || !KNOWN_PASSAGE_KINDS.has(kind)) {
          errs.push(`[kind] main passage ${ref} has unclassified kind ${JSON.stringify(kind)}`);
        }
      } else if (passage.kind !== undefined) {
        errs.push(`[kind] framing passage ${ref} must not carry kind (got ${JSON.stringify(passage.kind)})`);
      }
    }
  }
  return errs;
}

/**
 * Verify edition-kind coherence across the whole library (IDEA.md): every
 * edition carries a stamped `edition_kind` ("mula-only" | "commentarial"), a
 * mula-only edition carries a commentary in no part, and a commentarial
 * edition carries a commentary in at least one part (a uniform commentary drop
 * now fails against the committed expectation). A commentarial edition may
 * have an individual commentary-free part (e.g. a sarga whose whole text is
 * one un-glossed passage).
 *
 * @param libraryRoot - Absolute path to the library directory.
 * @returns List of error messages (empty when all editions are coherent).
 */
function checkEditionKindCoherence(libraryRoot: string): string[] {
  const errs: string[] = [];
  const hasCommentary = (obj: Record<string, unknown>): boolean => {
    const commentary = obj['commentary'];
    const commentaries = obj['commentaries'];
    if (commentary && typeof commentary === 'object' && Object.keys(commentary).length > 0) {
      return true;
    }
    return Array.isArray(commentaries) && commentaries.length > 0;
  };

  for (const filePath of collectJsonFiles(libraryRoot).sort()) {
    const data = JSON.parse(fs.readFileSync(filePath, 'utf-8')) as Record<string, unknown>;
    const kind = classifyFile(data);
    if (kind === 'edition-sub-envelope') {
      const editionId = data['edition_id'];
      const label = typeof editionId === 'string' ? editionId : path.basename(path.dirname(filePath));
      const stamp = data['edition_kind'];
      if (stamp !== 'mula-only' && stamp !== 'commentarial') {
        errs.push(`[edition_kind] edition ${label} missing/invalid edition_kind ${JSON.stringify(stamp)}`);
        continue;
      }
      const parts = data['parts'];
      if (!Array.isArray(parts)) continue;
      const dir = path.dirname(filePath);
      let anyPartHasCommentary = false;
      for (const part of parts as { file?: unknown }[]) {
        if (typeof part.file !== 'string') continue;
        const partPath = path.join(dir, part.file);
        if (!fs.existsSync(partPath)) continue;
        const partData = JSON.parse(fs.readFileSync(partPath, 'utf-8')) as Record<string, unknown>;
        const has = hasCommentary(partData);
        if (stamp === 'mula-only' && has) {
          errs.push(`[edition_kind] mula-only edition ${label} part ${part.file} has a commentary`);
        }
        anyPartHasCommentary = anyPartHasCommentary || has;
      }
      if (stamp === 'commentarial' && !anyPartHasCommentary) {
        errs.push(`[edition_kind] commentarial edition ${label} has commentary in no part`);
      }
    } else if (kind === 'grantha') {
      const label = (data['edition_id'] as string) ?? (data['grantha_id'] as string) ?? '?';
      const stamp = data['edition_kind'];
      if (stamp !== 'mula-only' && stamp !== 'commentarial') {
        errs.push(`[edition_kind] edition ${label} missing/invalid edition_kind ${JSON.stringify(stamp)}`);
        continue;
      }
      const expected = hasCommentary(data) ? 'commentarial' : 'mula-only';
      if (stamp !== expected) {
        errs.push(`[edition_kind] edition ${label} edition_kind ${stamp} mismatches commentary presence (${expected})`);
      }
    }
  }
  return errs;
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

const dataDir = path.resolve(__dirname, '../public/data/library');
const allFiles = collectJsonFiles(dataDir).sort();

let passCount = 0;
let failCount = 0;

for (const filePath of allFiles) {
  const rel = path.relative(dataDir, filePath);
  const data = JSON.parse(fs.readFileSync(filePath, 'utf-8')) as Record<string, unknown>;

  const kind = classifyFile(data);
  if (kind === null) continue;

  const validate = validatorForKind(kind);
  const schemaValid = validate(data);
  const schemaErrors: string[] = schemaValid
    ? []
    : (validate.errors ?? []).map(
        e => `[${e.instancePath || '(root)'}] ${e.message ?? ''}`
      );

  // Only run structure check for content files that have passages
  const structureErrors =
    kind === 'grantha' || kind === 'grantha-part'
      ? checkStructureConsistency(data)
      : [];

  // Edition paths in grantha-level envelopes must resolve under the library.
  const editionPathErrors =
    kind === 'grantha-envelope'
      ? checkEditionPathsResolve(data, dataDir)
      : [];

  // Per-block presentation: kind presence + classification (content files).
  const passageKindErrors =
    kind === 'grantha' || kind === 'grantha-part'
      ? checkPassageKinds(data)
      : [];

  const allErrors = [
    ...schemaErrors,
    ...structureErrors,
    ...editionPathErrors,
    ...passageKindErrors,
  ];

  if (allErrors.length > 0) {
    failCount++;
    console.error(`FAIL  ${rel}`);
    for (const msg of allErrors) {
      console.error(`      ${msg}`);
    }
  } else {
    passCount++;
  }
}

const total = passCount + failCount;
console.log(`\n=== ${passCount} PASS  ${failCount} FAIL  (${total} files scanned) ===`);

// Edition-kind coherence is a cross-file check (envelope vs its parts), run
// once over the whole library.
const editionKindErrors = checkEditionKindCoherence(dataDir);
for (const msg of editionKindErrors) {
  failCount++;
  console.error(`FAIL  ${msg}`);
}

if (failCount === 0) {
  console.log('All data files are valid.');
} else {
  process.exitCode = 1;
}

File: scripts/validate_data.py
from __future__ import annotations

import json
import pathlib
import warnings
from typing import Any
from jsonschema import Draft7Validator, RefResolver

warnings.filterwarnings('ignore', category=DeprecationWarning)


def _load_schemas(root: pathlib.Path) -> tuple[dict, dict]:
    """Load all root schema files and build a URI store for cross-file $ref resolution.

    Returns:
        A tuple of (schemas_by_key, store_by_uri) where store_by_uri is passed
        to RefResolver so that refs like grantha-envelope.schema.json#/definitions/...
        resolve correctly.
    """
    schema_paths = {
        'grantha':  root / 'grantha.schema.json',
        'envelope': root / 'grantha-envelope.schema.json',
        'part':     root / 'grantha-part.schema.json',
    }
    schemas = {k: json.loads(p.read_text()) for k, p in schema_paths.items()}
    store   = {p.as_uri(): schemas[k] for k, p in schema_paths.items()}
    return schemas, store, schema_paths


def _make_resolver(schema_key: str, schema_paths: dict, schemas: dict, store: dict) -> RefResolver:
    """Build a RefResolver anchored to the given schema file."""
    return RefResolver(
        base_uri=schema_paths[schema_key].as_uri(),
        referrer=schemas[schema_key],
        store=store,
    )


_VALID_KINDS: frozenset[str] = frozenset({
    'grantha-envelope',
    'edition-sub-envelope',
    'grantha-part',
    'grantha',
})

# Pinned classification (mirrors lib/data.ts KNOWN_PASSAGE_KINDS). Any kind
# found in the corpus without an entry here is a build error (per-block
# presentation model).
_KNOWN_PASSAGE_KINDS: frozenset[str] = frozenset({
    'Para', 'Gadya', 'Shloka', 'Mantra', 'Verse', 'Sutra',
})

_VALID_EDITION_KINDS: frozenset[str] = frozenset({'mula-only', 'commentarial'})


def _has_commentary(obj: dict[str, Any]) -> bool:
    """True when a part/edition carries a commentary (singular or plural)."""
    commentary = obj.get('commentary')
    commentaries = obj.get('commentaries')
    if isinstance(commentary, dict) and commentary:
        return True
    return isinstance(commentaries, list) and len(commentaries) > 0


def _check_passage_kinds(data: dict[str, Any]) -> list[str]:
    """Per-block kind invariants: main passages carry a classified `kind`;
    framing passages carry none. (Per-block presentation model.)"""
    errs: list[str] = []
    for key in ('passages', 'prefatory_material', 'concluding_material'):
        arr = data.get(key)
        if not isinstance(arr, list):
            continue
        for passage in arr:
            if not isinstance(passage, dict):
                continue
            ref = passage.get('ref', '?')
            if passage.get('passage_type') == 'main':
                kind = passage.get('kind')
                if kind not in _KNOWN_PASSAGE_KINDS:
                    errs.append(
                        f'[kind] main passage {ref} has unclassified kind {kind!r}'
                    )
            elif passage.get('kind') is not None:
                errs.append(
                    f'[kind] framing passage {ref} must not carry kind '
                    f'(got {passage.get("kind")!r})'
                )
    return errs


def _check_edition_kind_coherence(lib: pathlib.Path) -> list[str]:
    """Cross-file edition-kind coherence: every edition carries a stamped
    `edition_kind`, a mula-only edition has a commentary in no part, and a
    commentarial edition has a commentary in at least one part (a uniform drop
    now fails against the committed stamp). A commentarial edition may have an
    individual commentary-free part (e.g. a sarga whose whole text is one
    un-glossed passage). (Per-block presentation model.)"""
    errs: list[str] = []
    for path in sorted(lib.rglob('*.json')):
        data = json.loads(path.read_text())
        kind = _classify(data, path)
        if kind == 'edition-sub-envelope':
            label = data.get('edition_id') or path.parent.name
            stamp = data.get('edition_kind')
            if stamp not in _VALID_EDITION_KINDS:
                errs.append(
                    f'[edition_kind] edition {label} missing/invalid '
                    f'edition_kind {stamp!r}'
                )
                continue
            parts = data.get('parts')
            if not isinstance(parts, list):
                continue
            any_part_has_commentary = False
            for part in parts:
                if not isinstance(part, dict) or not part.get('file'):
                    continue
                part_path = path.parent / part['file']
                if not part_path.exists():
                    continue
                part_data = json.loads(part_path.read_text())
                has = _has_commentary(part_data)
                if stamp == 'mula-only' and has:
                    errs.append(
                        f'[edition_kind] mula-only edition {label} part '
                        f'{part["file"]} has a commentary'
                    )
                any_part_has_commentary = any_part_has_commentary or has
            if stamp == 'commentarial' and not any_part_has_commentary:
                errs.append(
                    f'[edition_kind] commentarial edition {label} has '
                    f'commentary in no part'
                )
        elif kind == 'grantha':
            label = data.get('edition_id') or data.get('grantha_id') or '?'
            stamp = data.get('edition_kind')
            if stamp not in _VALID_EDITION_KINDS:
                errs.append(
                    f'[edition_kind] edition {label} missing/invalid '
                    f'edition_kind {stamp!r}'
                )
                continue
            expected = 'commentarial' if _has_commentary(data) else 'mula-only'
            if stamp != expected:
                errs.append(
                    f'[edition_kind] edition {label} edition_kind {stamp} '
                    f'mismatches commentary presence ({expected})'
                )
    return errs


def _classify(data: dict[str, Any], path: pathlib.Path | None = None) -> str | None:
    """Return the file's explicit kind marker, or None to skip the file.

    Args:
        data: Parsed JSON content of the file.
        path: Unused; retained for call-site compatibility.

    Returns:
        The ``kind`` value if it is a recognised library file type, else None.
    """
    kind = data.get('kind')
    return kind if kind in _VALID_KINDS else None


def _validate_file(
    path: pathlib.Path,
    lib: pathlib.Path,
    schemas: dict,
    schema_paths: dict,
    store: dict,
) -> list[str]:
    """Validate a single file and return a list of error strings (empty = pass)."""
    try:
        data = json.loads(path.read_text())
    except json.JSONDecodeError as e:
        return [f'JSON parse error: {e}']

    kind = _classify(data, path)
    if kind is None:
        return []

    if kind == 'grantha-envelope':
        schema = schemas['envelope']
        resolver = _make_resolver('envelope', schema_paths, schemas, store)
    elif kind == 'edition-sub-envelope':
        schema = schemas['envelope']['definitions']['edition_sub_envelope']
        resolver = _make_resolver('envelope', schema_paths, schemas, store)
    elif kind == 'grantha-part':
        schema = schemas['part']
        resolver = _make_resolver('part', schema_paths, schemas, store)
    else:  # grantha
        schema = schemas['grantha']
        resolver = _make_resolver('grantha', schema_paths, schemas, store)

    validator = Draft7Validator(schema, resolver=resolver)
    errors = sorted(validator.iter_errors(data), key=lambda e: list(e.absolute_path))
    schema_msgs = [
        f'[{" > ".join(str(p) for p in e.absolute_path) or "(root)"}] {e.message[:150]}'
        for e in errors
    ]

    # Per-block presentation checks (content files only).
    if kind in ('grantha', 'grantha-part'):
        schema_msgs.extend(_check_passage_kinds(data))

    return schema_msgs


def main() -> None:
    """Validate all grantha data files in public/data/library/ against the root schemas."""
    root = pathlib.Path(__file__).parent.parent
    lib  = root / 'public' / 'data' / 'library'

    schemas, store, schema_paths = _load_schemas(root)

    pass_count = 0
    fail_count = 0

    for path in sorted(lib.rglob('*.json')):
        rel = str(path.relative_to(lib))
        errors = _validate_file(path, lib, schemas, schema_paths, store)
        if errors:
            fail_count += 1
            print(f'FAIL  {rel}')
            for msg in errors:
                print(f'      {msg}')
        else:
            pass_count += 1

    total = pass_count + fail_count
    print()
    print(f'=== {pass_count} PASS  {fail_count} FAIL  ({total} files scanned) ===')

    # Edition-kind coherence is a cross-file check, run once over the library.
    for msg in _check_edition_kind_coherence(lib):
        fail_count += 1
        print(f'FAIL  {msg}')

    if fail_count == 0:
        print('All data files are valid.')
    else:
        raise SystemExit(1)


if __name__ == '__main__':
    main()
